src/appleseed/foundation/image/oiioimagefilewriter.cpp

```diff
@@ -0,0 +1,515 @@
+{
+    assert(image_index < m_images->m_canvas.size());
+
+    // Retrieves canvas
+    assert(image_index < m_images->m_canvas.size());
+    const ICanvas* canvas = m_images->m_canvas[image_index];
+    assert(canvas);
+
+    // Retrieves canvas properties
+    const CanvasProperties& props = canvas->properties();
+
+    // Constructs the temporary buffer holding one row of tiles in target format.
+    std::unique_ptr<uint8> buffer(new uint8[props.m_canvas_width * props.m_tile_height * props.m_pixel_size]);
+    uint8* __restrict buffer_ptr = buffer.get();
+
+    // Loops over the rows of tiles
+    for (size_t tile_y = 0; tile_y < props.m_tile_count_y; tile_y++)
+    {
+        // Loops over the columns of tiles
+        for (size_t tile_x = 0; tile_x < props.m_tile_count_x; tile_x++)
+        {
+            // Retrieves the (tile_x, tile_y) tile.
+            const Tile& tile = canvas->tile(tile_x, tile_y);
+
+            // Loops over the row pixels of the current tile
+            for (size_t y = 0; y < tile.get_height(); y++)
+            {
+                // Loops over the column pixels of the current tile
+                for (size_t x = 0; x < tile.get_width(); x++)
+                {
+                    // Horizontal coordinate of the pixel in the temporary buffer.
+                    const size_t buffer_x = tile_x * props.m_tile_width + x;
+
+                    // Index of the pixel in the temporary buffer.
+                    const size_t buffer_index = (y * props.m_canvas_width + buffer_x) * props.m_pixel_size;
+
+                    // Retrieves the (x, y) pixel.
+                    const uint8* __restrict pixel = tile.pixel(x, y);
+
+                    // Write the pixel into the buffer.
+                    memcpy(&(buffer_ptr[buffer_index]), pixel, props.m_pixel_size);
+                }
+            }
+        }
+
+        // Computes y dimensional scanline border.
+        const size_t y_begin = tile_y;
+        const size_t y_end = y_begin + props.m_tile_height;
+
+        // Write scanline into the file.
+        if (!m_writer->write_scanlines(
+                static_cast<int>(y_begin), 
+                static_cast<int>(y_end), 
+                0, 
+                convert_pixel_format(props.m_pixel_format),
+                buffer_ptr))
+        {
+            const std::string msg = m_writer->geterror();
+            close_file();
+            throw ExceptionIOError(msg.c_str());
+        }
+    }
+}
+
+void OIIOImageFileWriter::write(const size_t image_index)
+{
+    assert(image_index < m_images->m_canvas.size());
+
+    if (m_writer->supports("tiles"))
+        write_tiles(image_index);
+    else
+        write_scanlines(image_index);
+}
+
+void OIIOImageFileWriter::write_single_image()
+{
+    if (!m_writer->open(m_filename, m_images->m_spec.back()))
+    {
+        const std::string msg = m_writer->geterror();
+        throw ExceptionIOError(msg.c_str());
```

Comment:
Here throwing exception makes perfect sense.

Reply:
I don't understand. It's an exception...